feat: Add K3s module

adar2 opened the pull request:
## What does this PR do?

Introduce K3s Module.

## Why is it important?

Allows testing k8s integration without using a real cluster.

adar2 committed:
added k3s container

HofmeisterAn committed:
chore: Clean up code smells

HofmeisterAn commented on src/Testcontainers.K3s/K3sBuilder.cs:
```diff
@@ -0,0 +1,101 @@
+    {
+        return base.Init().WithImage(RancherImage)
+            .WithPortBinding(KubeSecurePort)
+            .WithPortBinding(RancherWebhookPort)
```
Please use random assigned host ports `WithPortBinding(ushort, true)`. See our list of [best practices](https://dotnet.testcontainers.org/api/best_practices/). You can resolve the random assigned host port with `GetMappedPublicPort(ushort)`. I guess we need to considere this in getting the Kubeconfig File too.

HofmeisterAn commented on src/Testcontainers.K3s/K3sBuilder.cs:
```diff
@@ -0,0 +1,101 @@
+                    throw new TimeoutException();
+                }
+            }
+        }
```
You can replace the custom wait strategy simply with `Wait.ForUnixContainer().UntilMessageIsLogged(".*Node controller sync successful.*")`. If you need a lager regular expression timeout you can pass `new Regex("...", RegexOptions.None, TimeSpan.FromSeconds(60))` to it.

Or use something like [this](https://github.com/testcontainers/testcontainers-dotnet/blob/581a3cd85cb2b9b33b2550b6a82bf4b3913677ce/src/Testcontainers.Oracle/OracleBuilder.cs#L125-L136). This will not use a regular expression at all and is something I usually favor. Your readiness check string does not require a regular expression.

HofmeisterAn commented on src/Testcontainers.K3s/K3sContainer.cs:
```diff
@@ -0,0 +1,27 @@
+[PublicAPI]
+public sealed class K3sContainer : DockerContainer
+{
+    public string KubeConfigYaml { get; private set; }
```
Can we please replace this with something like `GetKubeconfigAsync()`?

HofmeisterAn committed:
chore: Clean up code smells

HofmeisterAn reviewed (changes requested):
Thanks for your pull request! Most parts look good, although I have a few change requests in regards of Testcontainers best practices. Happy to merge the PR soon as they are addressed 🙏.

adar2 committed:
using random assigned ports.
removed custom wait strategy.
removed KubeConfigYaml property.
introduced GetKubeConfig method.

HofmeisterAn committed:
chore: Clean up K3s module

HofmeisterAn commented on src/Testcontainers.K3s/K3sContainer.cs:
```diff
@@ -0,0 +1,32 @@
+    /// Gets the Kubeconfig.
+    /// </summary>
+    /// <returns>Task that completes when the Kubeconfig has been read.</returns>
+    public async Task<string> GetKubeconfigAsync()
+    {
+        var kubeconfigBytes = await ReadFileAsync("/etc/rancher/k3s/k3s.yaml")
+            .ConfigureAwait(false);
+
+        var kubeconfig = Encoding.Default.GetString(kubeconfigBytes);
+
+        var server = new UriBuilder(Uri.UriSchemeHttps, Hostname, GetMappedPublicPort(K3sBuilder.KubeSecurePort)).ToString();
+
+        return Regex.Replace(kubeconfig, "server:\\s?[:/\\.\\d\\w]+", "server: " + server, RegexOptions.None, TimeSpan.FromSeconds(1));
+    }
```
@adar2 I changed getting the Kubeconfig file to the code above. Please let me know what do you think (I adjusted the test too). I would like to avoid blocking the async context and remove the 3rd party dependency. Although the intention is nice, we try not to rely on 3rd party dependencies yet to avoid dependency clashes. I think modifying that small part works without YamlDotNet. If you are fine with the latests changes I am happy to merge the PR.
